Async chapter

sakex opened the pull request:
Async control flow + Async channels

sakex commented on src/async/control-flow.md:
```diff
@@ -0,0 +1,200 @@
+    let mut performances: Vec<AnimalRacePerformance> = Vec::new();
+    let timeout_sleep = sleep(timeout);
+    // See the chapter about pin.
+    tokio::pin!(timeout_sleep);
```
Alternatively, that can be done with a oneshot channel, so we don't need to talk about pin

djmitche replied:
Let's move this to the "pitfalls" section. AIUI the issue here is that you have a single Future that you'd like to `await` until it is Ready. This is different from the futures returned from `.recv()`, which can be created fresh every time around the loop. Maybe the example (in the pitfalls section) can start with
```rust
loop {
  select! {
    ..
    _ = sleep(timeout) => break,
  }
}
```
and then understand that this starts the timeout over for every iteration of the loop, and work out how to solve it from there.

I have to be honest, I don't think I could explain why `tokio::pin` is necessary here without a good bit of refreshing myself!

djmitche commented on src/async/control-flow.md:
````diff
@@ -0,0 +1,200 @@
+        }
+    );
+}
+```
````
This example seems a bit lengthy. Could this start with a shorter, simpler example with ideas in the speaker notes to improve it?

djmitche commented on src/async/pitfalls.md:
```diff
@@ -0,0 +1,79 @@
+# Pitfalls of async/await
```
Let's make each pitfall its own sub-chapter.

djmitche reviewed (commented):
Looking good! I think we want to keep each chapter pretty short, and at least initially the code examples should be short as well.

sakex commented on src/async/pitfalls.md:
````diff
@@ -0,0 +1,79 @@
+```
+
+Using `std::sync::mutex` in an `async` runtime is also very dangerous. When you lock the mutex in a thread then yield the executor using `.await` the thread might try to lock the mutex once more in a different task.
+Hence, prefer `async` alternatives like `tokio::sync::mutex`.
````
It's actually more of a `footgun`. I had a bad experience with this in the past. I will address your comments tomorrow, thank you :)

sakex commented:
@djmitche I addressed your comments and added a section about `daemon` tasks. Please have a look. I will start working on my exercise later this week!

djmitche commented on src/async/pitfalls/reentering-threads.md:
```diff
@@ -0,0 +1,51 @@
+# Reentering threads
```
I'm not sure how this is "reentering" - perhaps there's a better title?

djmitche reviewed (approved):
This looks good! I'll get it merged and we can continue to adjust in the issue181 branch.
